### ISSUE TITLE
Support state machines with the NF

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#8162]
Modelica state machines are still not supported by the NF. We should port the OF implementation to the NF.

**Comment by FWuellhorst** [comment]
What is the status of this, and is this related to https://github.com/OpenModelica/OpenModelica/issues/5694?

The following example does not check in OM for me, even though it's from a presentation about state machines in OM (https://openmodelica.org/images/docs/openmodelica2015/OpenModelica2015-talk14-OMStateMachines_Bernhard%20Thiele.pdf)

```
model Simple_NoAnnotations "Simple state machine"
  inner Integer i(start= 0);
  block State1
    outer output Integer i;
    output Integer j(start= 10);
  equation 
    i = previous(i) + 2;
    j = previous(j) - 1;
  end State1;
  State1 state1;
  block State2
   outer output Integer i;
  equation 
    i = previous(i) - 1;
  end State2;
  State2 state2;
equation 
  transition(state1, state2, i > 10, immediate= false, reset=true,  synchronize= false, priority= 1);
  transition(state2, state1, i < 1, immediate= false, reset=true,  synchronize= false,  priority= 1);
  initialState(state1);
end Simple_NoAnnotations;
```

Error:

```
[Simple_NoAnnotations: 18:3-18:101]: No matching function found for transition(/*State1*/ state1, /*State2*/ state2, /*Boolean*/ i > 10, immediate = /*Boolean*/ false, reset = /*Boolean*/ true, synchronize = /*Boolean*/ false, priority = /*Integer*/ 1).
Candidates are:
  transition() => ()
```

Version: OpenModelica v1.19.2 (64-bit)

**Comment by perost** [comment]
> What is the status of this, and is this related to #5694?

The status is that there's not been any work on this yet, it keeps being pushed down on my to do-list by other things. So it'll be done eventually, but I can't really say when.

I closed #5694 as a duplicate. This issue is technically the duplicate but the older one is on our Trac that we don't use anymore and didn't really contain anything important anyway.

**Comment by casella** [comment]
@FWuellhorst we currently have very high priority on #8425, that requires substantial frontend work and will bring huge benefits to everybody. As soon as that is completed, I think  @perost can start working on state machine support.

**Comment by FWuellhorst** [comment]
@casella @perost : Thanks for the quick response and all your work! The new features sound amazing! We will be happy to test them.

**Comment by FWuellhorst** [comment]
@casella: I will take this point for internal discussion. Currently, the initiative to provide OM support for our models is limited to few PhD students.

**Comment by casella** [comment]
@FWuellhorst is this still relevant for you? I understand @sturmk is also interested. If so, we can try to come up with a plan to finally port state machines in the NF as well.

**Comment by FWuellhorst** [comment]
@casella Yes, would be a nice feature, as a central model to calculate heating and cooling demands of air handling units does not run in OM. Also, I had a use-case for a state machine in a heat pump model developed for IBPSA. I used a fuzzy logic approach to enable OM compatibility, although a state machine would be better.

**Comment by casella** [comment]
At the moment @perost is looking into expandable connectors (#13067), which will probably take a few weeks. After that, we can consider porting state machine support from the OF to the NF.

@perost what do you think about that?

**Comment by casella** [comment]
@perost, based on the discussion at the Annual Meeting yesterday, this should be preferably done in 2025. Please take that into account in your planning.

**Comment by casella** [comment]
@sjoelund is trying to use Claude Code to get this done, see PR #15424. The port works with the C runtime, not with the C++ runtime, @rfranke is looking into that. Overall, it seems a promising approach. Stay tuned.

**Comment by pierre-haessig** [comment]
My introductory Modelica course at CentraleSupélec is getting rebooted next September in a new, shorter format, along with a SysML course. To make the link with SysML and other courses, I *may* be interested in simulating state machines, if available in OpenModelica.

Just to be sure I have a clear picture of the situation: what was the amount of support brought by the Old Frontend that @rfranke  is trying to port to the New Frontend?
- Modelica language (text only) support as described in the 2012 [State Machines in Modelica](https://ep.liu.se/en/conference-article.aspx?series=ecp&issue=76&Article_No=3) article, e.g. with the `transition` and `initialState` function support?
- graphical support (state machine schematic) as well?

**Comment by casella** [comment]
The porting is done by @perost, not @rfranke.

For sure Modelica language support, our current plan is to have it tentatively done by Claude code and the reviewed by @perost. Given the limited extent of the thing, this should work fine.

About graphical support, I'm not sure. @adeas31 I guess we need to add this stuff to the inst-based interface, right?

**Comment by perost** [comment]
> The porting is done by [@perost](https://github.com/perost), not [@rfranke](https://github.com/rfranke).

No, the porting was done by @sjoelund.

**Comment by adeas31** [comment]
> About graphical support, I'm not sure. [@adeas31](https://github.com/adeas31) I guess we need to add this stuff to the inst-based interface, right?

We have GUI support for state machines. If its just ported to the NF without any big changes then GUI should be fine.

**Comment by casella** [comment]
> > The porting is done by [@perost](https://github.com/perost), not [@rfranke](https://github.com/rfranke).
> 
> No, the porting was done by [@sjoelund](https://github.com/sjoelund).

Do we already have a PR for that?

**Comment by adrpo** [comment]
Already merged since some time now:
https://github.com/OpenModelica/OpenModelica/pull/15424

**Comment by casella** [comment]
So, can it be tested already?

**Comment by adrpo** [comment]
Yes, of course.

**Comment by casella** [comment]
@FWuellhorst, @pierre-haessig please proceed and report. Thanks!

**Comment by perost** [comment]
Are there any libraries that actually use state machines that we can test? The only thing I found was [AixLib.Airflow.AirHandlingUnit.Examples.AHU](https://libraries.openmodelica.org/branches/master/AixLib/files/AixLib_AixLib.Airflow.AirHandlingUnit.Examples.AHU.err), but that model seems to fail for compliance reasons unrelated to state machines.

**Comment by pierre-haessig** [comment]
> Are there any libraries that actually use state machines that we can test?

Same question: is there a readily available model for state machine? I just received the `1.27.0~dev.beta.1-4-ge5d8071-1` update (beta channel I guess) so that I guess that I can test it now.

**Comment by pierre-haessig** [comment]
> Same question: is there a readily available model for state machine? I just received the `1.27.0~dev.beta.1-4-ge5d8071-1` update (beta channel I guess) so that I guess that I can test it now.

Answering my own question: on `1.27.0~dev.beta.1-4-ge5d8071-1 `, I was able to run the StateMachine1 model from the “State Machines in Modelica” 2012 Modelica Conf article by Elmqvist et al. No annotation were provided, so I don't know what is the status of graphical support, since Diagram view is empty. 

Code:

```Modelica
model StateMachine1 "StateMachine1 model from the “State Machines in Modelica” 2012 Modeleca Conf article by Elmqvist et al. No annotation. PH 2026"

inner Integer i(start=0);

block State1
  outer output Integer i;
  equation
  i = previous(i) + 2;
end State1;

State1 state1;

block State2
  outer output Integer i;
  equation
  i = previous(i) - 1;
end State2;

State2 state2;

equation

initialState(state1);
transition(state1, state2, i > 10, immediate=false);
transition(state2, state1, i < 1, immediate=false);

end StateMachine1;
```

(to be simulated on 20 seconds, since for some reason there is a default Clock ticking every second, not clear in the article)

**Comment by casella** [comment]
I guess that's the same example that is reported in the MLS, [Section 17.3.7](https://specification.modelica.org/maint/3.7/state-machines.html#example).

I googled a bit and I found this [tutorial](https://2019.international.conference.modelica.org/proceedings/html/tutorials/Tutorial_07_Abstract.pdf). @gkurzbach, @TorstenBlochwitz, do you still have the source code for that and can you share it?

**Comment by TorstenBlochwitz** [comment]
> I guess that's the same example that is reported in the MLS, [Section 17.3.7](https://specification.modelica.org/maint/3.7/state-machines.html#example).
> 
> I googled a bit and I found this [tutorial](https://2019.international.conference.modelica.org/proceedings/html/tutorials/Tutorial_07_Abstract.pdf). [@gkurzbach](https://github.com/gkurzbach), [@TorstenBlochwitz](https://github.com/TorstenBlochwitz), do you still have the source code for that and can you share it?

Hi Francesco,
unfortunately, we do'nt have anything about that tutorial. There have been too many IT-changes since then.

**Comment by FWuellhorst** [comment]
Not at the institute maintaining AixLib anymore, so no access to Dymola. But checked with current main and fixed some minor issues. 

Getting the following errors for various variables in the AHU model:

[77] 10:30:18 Übersetzung Fehler
Following variable is discrete, but does not appear on the LHS of a when-statement: 'multizone.AirHandlingUnit.onlyHeatingHRS_true.phi_t'.

branch: https://github.com/RWTH-EBC/AixLib/tree/om_state_machine_test

Not sure why this happens or how to fix it, any idea?
Never implemented state machines myself and outside of that model we don't use them, sadly. But nice to see support in OM for this!!

**Comment by casella** [comment]
@AnHeuermann mind having a quick look?

**Comment by casella** [comment]
@FWuellhorst I'm closing this ticket to document the new feature in the release notes of 1.27.0. Please open other tickets on any issue you may encounter, including [this one](https://github.com/OpenModelica/OpenModelica/issues/8162#issuecomment-4852217432).
